src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4288,14 +4288,173 @@ private static bool UseLegacyWarnings(BoundExpression expr)
+
+                for (int i = 0; i < n; i++)
+                {
+                    var variable = variables[i];
+                    var rightPart = rightParts[i];
+                    var nestedVariables = variable.NestedVariables;
+                    if (nestedVariables != null)
+                    {
+                        VisitDeconstructionArguments(nestedVariables, conversion.UnderlyingConversions[i], rightPart);
+                    }
+                    else
+                    {
+                        int targetSlot = MakeSlot(variable.Expression);
+                        var targetType = variable.Type;
+                        var valueType = VisitOptionalImplicitConversion(rightPart, targetType, useLegacyWarnings: true, AssignmentKind.Assignment);
+                        int valueSlot = MakeSlot(rightPart);
+                        TrackNullableStateForAssignment(rightPart, targetType, targetSlot, valueType, valueSlot);
+                    }
+                }
+            }
+        }
+
+        private struct DeconstructionVariable
+        {
+            internal readonly BoundExpression Expression;
+            internal readonly TypeSymbolWithAnnotations Type;
+            internal readonly ArrayBuilder<DeconstructionVariable> NestedVariables;
+
+            internal DeconstructionVariable(BoundExpression expression, TypeSymbolWithAnnotations type)
+            {
+                Expression = expression;
+                Type = type;
+                NestedVariables = null;
+            }
+
+            internal DeconstructionVariable(ArrayBuilder<DeconstructionVariable> nestedVariables)
+            {
+                Expression = null;
+                Type = default;
+                NestedVariables = nestedVariables;
+            }
+        }
+
+        private ArrayBuilder<DeconstructionVariable> GetDeconstructionAssignmentVariables(BoundTupleExpression expr)
+        {
+            var arguments = expr.Arguments;
+            var builder = ArrayBuilder<DeconstructionVariable>.GetInstance(arguments.Length);
+            foreach (var argument in arguments)
+            {
+                builder.Add(GetDeconstructionAssignmentVariable(argument));
+            }
+            return builder;
+        }
+
+        private DeconstructionVariable GetDeconstructionAssignmentVariable(BoundExpression expr)
+        {
+            switch (expr.Kind)
+            {
+                case BoundKind.TupleLiteral:
+                case BoundKind.ConvertedTupleLiteral:
+                    return new DeconstructionVariable(GetDeconstructionAssignmentVariables((BoundTupleExpression)expr));
+                default:
+                    VisitLvalue(expr);
+                    return new DeconstructionVariable(expr, _resultType);
+            }
+        }
+
+        // cf. LocalRewriter.GetRightParts.
+        private ImmutableArray<BoundExpression> GetDeconstructionRightParts(BoundExpression expr)
```

Comment:
Local function?

Reply:
Will leave as is, so there are top-level methods for getting the parts of the LHS and the parts of the RHS.

---
In reply to: [252749367](https://github.com/dotnet/roslyn/pull/32994#discussion_r252749367) [](ancestors = 252749367)